text/0139-faster-erasure-coding.md

```diff
@@ -0,0 +1,222 @@
+	/// The index of this erasure-encoded chunk of data.
+	pub index: u16,
+	/// Proof for this chunk against an erasure root.
+	pub proof: Proof,
+}
+
+impl Iterator for MerklizedChunks {
+	type Item = ErasureChunk;
+
+	fn next(&mut self) -> Option<Self::Item> {
+		let chunk = self.data.pop_front()?;
+		let d = self.tree.len() - 1;
+		let idx = self.current_index.0;
+		let mut index = idx as usize;
+		let mut path = Vec::with_capacity(d);
+		for i in 0..d {
+			let layer = &self.tree[i];
+			if index % 2 == 0 {
+				path.push(layer[index + 1]);
+			} else {
+				path.push(layer[index - 1]);
+			}
+			index /= 2;
+		}
+		self.current_index += 1;
+		Some(ErasureChunk {
+			chunk,
+			proof: Proof::try_from(path).expect("the path is limited by tree depth; qed"),
+			index: idx,
+		})
+	}
+}
+
+impl MerklizedChunks {
+	/// Compute `MerklizedChunks` from a list of erasure chunks.
+	pub fn compute(chunks: Vec<Vec<u8>>) -> Self {
+		let mut hashes: Vec<Hash> = chunks
+			.iter()
+			.map(|chunk| {
+				let hash = hash_fn(chunk);
+				Hash::from(hash)
+			})
+			.collect();
+		hashes.resize(chunks.len().next_power_of_two(), Hash::default());
+
+		let depth = hashes.len().ilog2() as usize + 1;
+		let mut tree = vec![Vec::new(); depth];
+		tree[0] = hashes;
+
+		// Build the tree bottom-up.
+		(1..depth).for_each(|lvl| {
+			let len = 2usize.pow((depth - 1 - lvl) as u32);
+			tree[lvl].resize(len, Hash::default());
+
+			// NOTE: This can be parallelized.
+			(0..len).for_each(|i| {
+				let prev = &tree[lvl - 1];
+
+				let hash = combine(prev[2 * i], prev[2 * i + 1]);
+
+				tree[lvl][i] = hash;
+			});
+		});
+
+		assert!(tree[tree.len() - 1].len() == 1, "root must be a single hash");
+
+		Self {
+			root: ErasureRoot::from(tree[tree.len() - 1][0]),
+			data: chunks.into(),
+			tree,
+			current_index: 0,
+		}
+	}
+}
+
+fn combine(left: Hash, right: Hash) -> Hash {
+	let mut hasher = Hasher::new();
+
+	hasher.update(left.0.as_slice());
+	hasher.update(right.0.as_slice());
+
+	hasher.finalize().into()
+}
+
+impl ErasureChunk {
+	/// Verify the proof of the chunk against the erasure root and index.
+	pub fn verify(&self, root: &ErasureRoot) -> bool {
+		let leaf_hash = Hash::from(hash_fn(&self.chunk));
+		let bits = Bitfield(self.index.0);
+
+		let root_hash = self.proof.0.iter().fold((leaf_hash, 0), |(acc, i), hash| {
+			let (a, b) = if bits.get_bit(i) { (*hash, acc) } else { (acc, *hash) };
+			(combine(a, b), i + 1)
+		});
+
+		// check the index doesn't contain more bits than the proof length
+		let index_bits = 16 - self.index.0.leading_zeros() as usize;
+		index_bits <= self.proof.0.len() && root_hash.0 == root.0
+	}
+}
+
+struct Bitfield(u16);
+
+impl Bitfield {
+	/// Get the bit at the given index.
+	pub fn get_bit(&self, i: usize) -> bool {
+		self.0 & (1u16 << i) != 0
+	}
+}
+```
+
+## Drawbacks
+
+Bundling breaking changes with RFC 47 might reset the progress of updating collators. However, the omni node initiative can alleviate this problem.
```

Comment:
If we do have pressure on the DA, we can probably expedite the roll out by making a community announcement to upgrade their collators (easier with omninode) after we get this feature on all validators. Giving a 6 month notice period should be enough. 

It would be useful to know how many times PoVs are retrieved by collators, I expect this to be a very rare event.